FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -57,12 +73,198 @@ object PgpMsg {
+
+    val keyList: List<PGPSecretKeyRing>
+    try {
+      keyList = keys.map {
+        if (it.passphrase != null) {
+          PgpKey.decryptKey(it.keyRing, it.passphrase)
+        } else {
+          if (KeyRingInfo(it.keyRing).isFullyDecrypted) {
+            throw PGPException("flowcrypt: need passphrase")
+          }
+          it.keyRing
+        }
+      }.toList()
+    } catch (ex: PGPException) {
+      if (ex.message == "flowcrypt: need passphrase") {
+        return DecryptionResult.withError(
+          type = DecryptionErrorType.NEED_PASSPHRASE,
+          message = "Need passphrase"
+        )
+      }
+      return DecryptionResult.withError(
+        type = DecryptionErrorType.WRONG_PASSPHRASE,
+        message = "Wrong passphrase",
+        cause = ex
+      )
+    }
+
+    var exception: Exception? = null
+    try {
+      val d = PGPainless.decryptAndOrVerify()
+        .onInputStream(input)
+        .decryptWith(UnprotectedKeysProtector(), PGPSecretKeyRingCollection(keyList))
+
+      val decryptionStream = if (pgpPublicKeyRingCollection != null) {
+        d.verifyWith(pgpPublicKeyRingCollection).ignoreMissingPublicKeys().build()
+      } else {
+        d.doNotVerify().build()
+      }
+
+      val output = ByteArrayOutputStream()
+      decryptionStream.use { it.copyTo(output) }
+
+      val streamResult = decryptionStream.result
+      return DecryptionResult.withDecrypted(output, streamResult.fileInfo?.fileName)
+    } catch (ex: MessageNotIntegrityProtectedException) {
+      return DecryptionResult.withError(
+        type = DecryptionErrorType.NO_MDC,
+        message = "Security threat! Message is missing integrity checks (MDC)." +
+            " The sender should update their outdated software.",
+        cause = ex
+      )
+    } catch (ex: ModificationDetectionException) {
+      return DecryptionResult.withError(
+        type = DecryptionErrorType.BAD_MDC,
+        message = "Security threat! Integrity check failed.",
+        cause = ex
+      )
+    } catch (ex: PGPDataValidationException) {
+      return DecryptionResult.withError(
+        type = DecryptionErrorType.KEY_MISMATCH,
+        message = "There is no matching key",
+        cause = ex
+      )
+    } catch (ex: PGPException) {
+      if (
+        ex.message?.contains("exception decrypting session info") == true
+        || ex.message?.contains("encoded length out of range") == true
+        || ex.message?.contains("Exception recovering session info") == true
+      ) {
+        DecryptionResult.withError(
+          type = DecryptionErrorType.KEY_MISMATCH,
+          message = "There is no matching key",
+          cause = ex
+        )
+      } else {
+        exception = ex
+      }
+    } catch (ex: Exception) {
+      exception = ex
+    }
+    return DecryptionResult.withError(
+      type = DecryptionErrorType.OTHER,
+      message = "Decryption failed",
+      cause = exception
+    )
+  }
```

Comment:
Exception handling looks better now - over time we can fine tune it based on real world errors